include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,223 @@
+	struct torrent_entry
+	{
+		std::string name;
+		std::set<peer_entry> peers;
+	};
+
+	struct dht_immutable_item
+	{
+		dht_immutable_item() : value(0), num_announcers(0), size(0) {}
+		// malloced space for the actual value
+		char* value;
+		// this counts the number of IPs we have seen
+		// announcing this item, this is used to determine
+		// popularity if we reach the limit of items to store
+		bloom_filter<128> ips;
+		// the last time we heard about this
+		time_point last_seen;
+		// number of IPs in the bloom filter
+		int num_announcers;
+		// size of malloced space pointed to by value
+		int size;
+	};
+
+	struct ed25519_public_key { char bytes[item_pk_len]; };
+
+	struct dht_mutable_item : dht_immutable_item
+	{
+		char sig[item_sig_len];
+		boost::uint64_t seq;
+		ed25519_public_key key;
+		char* salt;
+		int salt_size;
+	};
+
+	struct count_peers
+	{
+		int* count;
+		count_peers(int* c): count(c) {}
+		void operator()(std::pair<libtorrent::sha1_hash
+				, libtorrent::dht::torrent_entry> const& t)
+		{
+			*count += t.second.peers.size();
+		}
+	};
+
+	// return true of the first argument is a better candidate for removal, i.e.
+	// less important to keep
+	struct immutable_item_comparator
+	{
+		immutable_item_comparator(sha1_hash const& our_id) : m_our_id(our_id) {}
+		immutable_item_comparator(immutable_item_comparator const& c)
+				: m_our_id(c.m_our_id) {}
+
+		bool operator() (std::pair<sha1_hash, dht_immutable_item> const& lhs
+				, std::pair<sha1_hash, dht_immutable_item> const& rhs) const
+		{
+			int l_distance = distance_exp(lhs.first, m_our_id);
+			int r_distance = distance_exp(rhs.first, m_our_id);
+
+			// this is a score taking the popularity (number of announcers) and the
+			// fit, in terms of distance from ideal storing node, into account.
+			// each additional 5 announcers is worth one extra bit in the distance.
+			// that is, an item with 10 announcers is allowed to be twice as far
+			// from another item with 5 announcers, from our node ID. Twice as far
+			// because it gets one more bit.
+			return lhs.second.num_announcers / 5 - l_distance < rhs.second.num_announcers / 5 - r_distance;
+		}
+
+	private:
+
+		// explicitly disallow assignment, to silence msvc warning
+		immutable_item_comparator& operator=(immutable_item_comparator const&);
+
+		node_id const& m_our_id;
+	};
+
+	// internal
+	inline bool operator<(ed25519_public_key const& lhs, ed25519_public_key const& rhs)
+	{
+		return memcmp(lhs.bytes, rhs.bytes, sizeof(lhs.bytes)) < 0;
+	}
+
+	// internal
+	inline bool operator<(peer_entry const& lhs, peer_entry const& rhs)
+	{
+		return lhs.addr.address() == rhs.addr.address()
+			   ? lhs.addr.port() < rhs.addr.port()
+			   : lhs.addr.address() < rhs.addr.address();
+	}
+
+	struct TORRENT_EXPORT dht_storage_interface
+	{
+		virtual size_t num_torrents() const = 0;
+		virtual size_t num_peers() const = 0;
+
+		virtual torrent_entry const* lookup_peers(sha1_hash const& info_hash) const = 0;
```

Comment:
returning a pointer here either enforces a specific implementation, or it requires an implementation to be less efficient by creating a temporary torrent_entry object and copy all peers into it (even though we're only interested in a subset of peers). I think the interface would be more flexible the function was expected to fill in a vector (or maybe just a buffer) of a random subset of peers instead. The most efficient interface would probably be to ask it fill in binary, compact IP-port pairs into a buffer. then it could be written directly into the packet